_posts/2020-02-05-keycloak-z-angularem-i-keycloakiem.md

```diff
@@ -0,0 +1,427 @@
+
+### Zależności
+
+Na backendzie wykorzystamy zależności:
+
+- `org.keycloak:keycloak-adapter-core`
+- `org.keycloak:keycloak-spring-boot-2-adapter`
+- `org.keycloak:keycloak-tomcat-adapter`
+
+Moglibyśmy wykorzystać jedną zależność, zawierającą powyższe adaptery:
+
+- `org.keycloak:keycloak-spring-boot-2-starter`
+
+jednak w najnowszej dostępnej wersji `4.0.0.Final` wykorzystuje stare wersje adapterów. Jedną ze zmian w nowszych wersjach adapterów które wykorzystamy, jest poprawiona walidacja tokenów. Jeśli w przyszłości pojawi się nowa wersja tej zależności, z nowymi wersjami adapterów, to nic nie stoi na przeszkodzie żeby ją wykorzystać.
+
+### Konfiguracja
+
+Oprócz konfiguracji portu aplikacji oraz poziomu logowania adapterów (co pozwoli nam lepiej śledzić w logach co się dzieje), musimy skonfigurować `keycloak-adapter-core`.  
+Dzięki wykorzystaniu `keycloak-spring-boot-2-adapter` możemy wszystko skonfigurować w `application.yml`:
+
+```yaml
+keycloakRequiredUserRole: user_role
+keycloak:
+  enabled: true
+  auth-server-url: http://localhost:8180/auth
+  realm: SpringBootAngular
+  resource: SpringBootAngularClient
+  security-constraints:
+    - authRoles:
+        - ${keycloakRequiredUserRole}
+      securityCollections:
+        - name: protected resource
+          patterns:
+            - /api/*
+    - securityCollections:
+        - name: public resource
+          patterns:
+            - /api/keycloak/config
+  credentials:
+    secret: "<SECRET>"
+  realm-key: "<PUBLIC_KEY>"
+```
+
+- `keycloakRequiredUserRole` - dla ułatwienia, ponieważ wykorzystamy tylko jedną rolę, będzie nam łatwiej udostępnić ją przez api (jeśli chcemy wykorzystać więcej ról, to musimy je wyciągąć z listy `authRoles`  
+- `keycloak.security-constraints` - tutaj definiujemy ograniczenia endpointów
+
+Możemy zdefiniować ścieżki dostępne publicznie:  
+np. `/api/keycloak/config`  
+oraz ścieżki które będą wymagały uprawnień  
+np. `/api/*`  
+które będzie wymagać od użytkownika roli `${keycloakRequiredUserRole}` (czyli `user_role`).
+
+**Konfiguracja Keycloak:**
+
+- `keycloak.enabled` - umożliwi nam łatwe wyłączenie uwierzytelniania  
+- `keycloak.auth-server-url` - adres naszego serwera Keycloak  
+- `keycloak.realm` - nazwa naszego realmu  
+- `keycloak.resource` - nazwa naszego klienta skonfigurowanego dla podanego realmu  
+- `keycloak.credentials.secret` - secret wygenerowany w `SpringBootAngularClient`, możemy go znaleźć w konsoli administracyjnej (`Clients > SpringBootAngularClient > Credentials > Secret`)  
+- `keycloak.realm-key` - klucz publiczny realmu, możemy go znaleźć w konsoli administracyjnej (`Realm Settings > Keys > Active > RSA > Public Key`)
+
+### Wystawienie konfiguracji dla frontendu
+
+Ponieważ zabezpieczamy osobno aplikację backendową jak i frontendową, a konfiguracja Keycloak jest taka sama, to możemy sobie wystawić konfigurację z `application.yml` przez api. Dzięki temu unikniemy duplikowania konfiguracji w aplikacji frontendowej.
```

Comment:
Moja propozycja jest taka:
Ponieważ osobno zabezpieczamy aplikację zarówno backendową jak i frontendową (a konfiguracja Keycloak jest taka sama), to możemy wystawić tę konfigurację z `application.yml` przez api.

Reply:
Można tak jak proponuje Adam, a w szczególności proponuję usunąć "sobie"